feat: Libfabric support

chandlj opened the pull request:
#### Overview:

This PR updates the Dynamo Dockerfiles to build NIXL with libfabric support and install the necessary runtime packages. Builds off #4317 to allow users to set the `NixlConnector` backend to `LIBFABRIC` using `--kv-transfer-config`.

#### Details:

Builds off #4317 by updating each respective framework Dockerfile as well as the base Dockerfile to build NIXL v0.7.1 with Libfabric support. Also bumps the vLLM version to 0.11.0 to have support for setting `NixlConnector` backends.

#### Where should the reviewer start?

`Dockerfile`, `Dockerfile.[framework]`, `pyproject.toml`

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

Relates to #4185 


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **Chores**
  * Updated NIXL dependency from 0.7.0 to 0.7.1
  * Updated vLLM dependency to 0.11.0
  * Updated sglang NIXL dependency to 0.7.1
  * Added libfabric v2.3.0 support across container images for enhanced networking capabilities

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

chandlj committed:
fix: Parse custom kv config

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

chandlj committed:
Revert "fix: Parse custom kv config" because not needed on vLLM 0.11.0

This reverts commit 1971265884c23b947821afd4b65ffd7651d2838e.

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

chandlj committed:
feat: Build NIXL with Libfabric support for AWS EFA

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

chandlj committed:
feat: Add Libfabric build arg to build script

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

chandlj committed:
Add libfabric to TRTLLM and SGLANG containers

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

chandlj committed:
Update libfabric install instructions

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

rmccorm4 commented:
Hi @chandlj, since #4317 was merged can you help pull in the latest updates and fix the merge conflicts?

chandlj commented:
> Hi @chandlj, since #4317 was merged can you help pull in the latest updates and fix the merge conflicts?

@rmccorm4 done!

nv-tusharma commented on container/Dockerfile:
```diff
@@ -270,6 +277,35 @@ RUN --mount=type=secret,id=aws-key-id,env=AWS_ACCESS_KEY_ID \
+    export CC=${CC} && \
+    export CXX=${CXX} && \
+    cd /usr/local/src && \
+    git clone https://github.com/ofiwg/libfabric.git && \
```
Got it, I think libfabric provides pre-built packages, https://github.com/ofiwg/libfabric?tab=readme-ov-file#installing-pre-built-libfabric-packages, would prefer if we could just pin to the dpkg package instead of building from source. The reason we do it for UCX & NIXL today is since the headers are required for dynamo integration.

kahalon replied:
This method (that we use in NIXL) was recommended one by AWS.

erezzarum replied:
Using the EFA installer will override other packages that will be optimized for runtime, you could extract the libfabric packages from the EFA installer for just using libfabric libraries, as currently the EFA installer doesn't have flags to just install libfabric libraries, you will also have to test if the provided packages will satisfy manylinux, if not, it's ok to build against upstream libfabric as we do in NIXL.

erezzarum commented:
I was about to create an identical PR (AWS side) with the same approach of compiling the upstream libfabric against the NIXL library as we do the same in the NIXL repo.

For the runtime image, we do recommend to use our EFA installer and use the libraries that come with it, but after this is merged, we could use the upstream NGC Dynamo containers and use the EFA installer there which will simplify using Dynamo on AWS, this is identical to how our vLLM DLC is built, you can look [here](https://github.com/aws/deep-learning-containers/blob/master/vllm/x86_64/gpu/Dockerfile)

Thanks!

chandlj commented:
> @chandlj - Are you planning on moving forward with this PR?

@dagil-nvidia Was waiting for next steps from you. From my perspective, it seems like this is ready. @erezzarum said he would follow up on this with an EFA-specific container, but I can also tackle if needed. I can also go ahead and attempt to resolve the merge conflicts.

erezzarum commented on container/Dockerfile.sglang:
```diff
@@ -326,6 +328,7 @@ ${NIXL_LIB_DIR}:\
 ${NIXL_PLUGIN_DIR}:\
 /usr/local/ucx/lib:\
 /usr/local/ucx/lib/ucx:\
+/usr/local/libfabric/lib:\
```
No need for that, don't copy the libfabric built from the base container

chandlj committed:
fix: Remove libfabric from runtime images

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

chandlj committed:
feat: Add AWS-specific stages for framework images (runtime-aws, dev-aws)

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

chandlj committed:
feat: Add option for --target local-dev-aws in addition to local-dev

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

chandlj committed:
fix: remove extra comment about libfabric

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

chandlj committed:
fix: Remove redeclared build args

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

chandlj committed:
fix: Change install to be more like review example

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

chandlj committed:
fix: Address comments, simplify dev-aws stages

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

chandlj commented:
@erezzarum Addressed your comments, left a comment about the `ENTRYPOINT` and `CMD`, not sure if you still think I should change it but I did it to match the previous stages.

@dagil-nvidia Thanks for the heads up. I believe I bumped vLLM to 0.12.0 in this to match, and sounds good on the NIXL 0.8.0. If that gets updated before we merge this, I'll make sure to change it

erezzarum commented on container/Dockerfile.sglang:
```diff
@@ -529,3 +529,151 @@ RUN python3 -m pip install --no-cache-dir \
+USER dynamo
+
+ENTRYPOINT ["/opt/nvidia/nvidia_entrypoint.sh"]
+CMD []
```
Would you be able to test it?

chandlj committed:
fix: Declare EFA_VERSION at top once, reuse later

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

chandlj commented on container/Dockerfile.sglang:
```diff
@@ -529,3 +529,70 @@ RUN python3 -m pip install --no-cache-dir \
+
+FROM runtime AS runtime-aws
+
+ARG EFA_VERSION=1.45.1
```
Good catch, fixed

erezzarum commented:
@chandlj Are you able to test it?

erezzarum commented:
Confirmed
`./container/build.sh --target runtime-aws`
Build a vLLM runtime with EFA installer, tested disaggregated serving on p5.48xlarge using LIBFABRIC backend
Also tested NIXL 0.8.0

nnshah1 commented on container/Dockerfile.sglang:
```diff
@@ -529,3 +531,70 @@ RUN python3 -m pip install --no-cache-dir \
+ENTRYPOINT ["/opt/nvidia/nvidia_entrypoint.sh"]
+CMD []
+
+########################################################################
```
as a thought - is it possible to have a common aws dockerfile that would take any runtime image as base and make the changes? in order to not have to change the files directly here?

chandlj committed:
fix: Remove unnecessary libfabric line in LD_LIBRARY_PATH

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

chandlj committed:
fix: Simplify AWS container by adding dedicated Dockerfile.aws

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

erezzarum commented:
@chandlj can you please rebase/merge with the new changes that were made in main, there was a major refactor, to reduce time during build/push, currently it shifted to have all the dynamo_base stage in the runtime Dockerfile, so there's duplicate code for dynamo_base per each runtime and it no longer relies on the base image.

chandlj commented:
@erezzarum Done!

erezzarum commented:
> @erezzarum Done!

Please note the major change, the dynamo_base is now being build inside each runtime, so you will have to duplicate what you have done in base across all runtimes (vllm, trtllm, sglang)

chandlj committed:
fix: Replicate libfabric install step in all framework images

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

chandlj committed:
fix: buld.sh local-dev duplicated step

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

chandlj committed:
fix: Use proper sccache configuration

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

erezzarum commented on container/Dockerfile.vllm:
```diff
@@ -272,6 +276,36 @@ RUN --mount=type=secret,id=aws-key-id,env=AWS_ACCESS_KEY_ID \
      ldconfig
 
+ARG NIXL_LIBFABRIC_REF
+RUN --mount=type=secret,id=aws-key-id,env=AWS_ACCESS_KEY_ID \
```
Thanks, i will try to run through this on Sunday

chandlj committed:
fix: buildkit bug

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

erezzarum commented:
> @erezzarum I think my most recent change just fixed it, can you try again?

Can you add the `$PLATFORM` as well?

chandlj commented:
> Can you add the `$PLATFORM` as well?

Sorry, I'm not following where this is used in the AWS image, can you explain a little more? Is the EFA installer ARCH-specific?

chandlj committed:
fix: Add platform to AWS image build

Signed-off-by: Joe Chandler <jchandler@glia-ai.com>

nnshah1 commented on container/build.sh:
```diff
@@ -455,13 +466,14 @@ show_help() {
     echo "  [--build-context name=path to add build context]"
     echo "  [--release-build perform a release build]"
-    echo "  [--make-efa Enables EFA support for NIXL]"
+    echo "  [--make-efa Adds AWS EFA layer on top of the built image (works with any target)]"
```
nit:  Non blocking - but 'enable' or 'add' seems a little more intuitive - but again not a blocker

```suggestion
    echo "  [--enable-efa Adds AWS EFA layer on top of the built image (works with any target)]"
```

nvda-mesharma committed:
Merge branch 'main' into libfabric-support

nvda-mesharma committed:
Merge branch 'main' into libfabric-support
